src/core/thread/network_data_tlvs.cpp

```diff
@@ -0,0 +1,120 @@
+/*
+ *  Copyright (c) 2016-21, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements methods for generating and processing Thread Network Data TLVs.
+ */
+
+#include "network_data_tlvs.hpp"
+
+namespace ot {
+namespace NetworkData {
+
+//---------------------------------------------------------------------------------------------------------------------
+// NetworkDataTlv
+
+const NetworkDataTlv *NetworkDataTlv::Find(const NetworkDataTlv *aStart, const NetworkDataTlv *aEnd, Type aType)
+{
+    const NetworkDataTlv *tlv;
+
+    for (tlv = aStart; (tlv + 1 <= aEnd) && (tlv->GetNext() <= aEnd); tlv = tlv->GetNext())
```

Comment:
```suggestion
    for (tlv = aStart; (tlv < aEnd) && (tlv->GetNext() <= aEnd); tlv = tlv->GetNext())
```

Here `tlv + 1` moves the address of `tlv` forward by 4 bytes (pointer size). What's the point of comparing that address with `aEnd`?

Reply:
Note that this is just an existing implementation (just moved from `NetworkData` class to `NetworkDataTlv` class)

Also the `(tlv + 1 <= aEnd)` is very much intentional. We use this pattern in a lot of places when dealing with `Tlv`s:

- `tlv + 1` increases the pointer by the size of its type, which `NetworkDataTlv`. 
- `NetworkDataTlv` stuct contains two `uint8_t` member vairables: `mType` and `mLength`.
- So `tlv + 1` points to end of `NetworkDataTlv` struct (or two bytes after the `tlv` address).
- The  `(tlv + 1 <= aEnd)` checks that the type and length fields are before `aEnd` pointer.
- This check is important before the `tlv->GetNext()` call which itself will access `mLength` field (otherwise we may read beyond `aEnd` pointer from `tlv->GetNext()` call which is not generally safe and also results in warnings/errors by fuzzer and/or other analyzer tools).

Reply:
Oh I made a mistake here. Thanks.